docs/features/component-versioning.md

```diff
@@ -0,0 +1,88 @@
+# Roslyn Component Versioning
+
+This document proposes a scheme that allows the authors of Roslyn Components (Analyzers, Generators etc.) to specify the
+set of features the compiler is required to provide in order to load the specified component.
+
+## Rationale
+
+Today, as an author of a Roslyn component, there is no way to declare a particular version of roslyn that you depend on. For instance
+a Generator that requires the _PostInitialization_ feature will fail to load in compilers that did not support it, but only by
+virtue of the roslyn version it references containing types not available in the older compiler.
+
+With analyzers this was generally less of a problem, as the API shape tended to be more stable, and Syntax is generally additive. Newer
+analyzers could continue to work in older compilers as the code paths referencing new syntax types would not be hit. However, newer APIs, such as those introduced for nullable analysis, still have the load time failure issue.
+
+Although blocked from loading, the current user experience of using a component targeting a newer compiler version is sub-par and opaque as to the action the
+user needs to take to resolve it.
+
+Instead, it would be ideal if the author had a way to specify a particular feature of the compiler that is required; older compilers (at least the
+ones going forward from now on) would be able to detect this feature requirement at load time, and specify a specific warning that this version of the compiler cannot load the component.
+
+## Proposal
+
+We introduce a new Attribute `[RoslynFeature(...)]` that a component author can apply to their component definition. The attribute would take a required string parameter, which is a comma separated list of features the component requires in order to load.
+
+`[RoslynFeature("Generator.PostInitialization, LanguageVersion9")]`
```

Comment:
If we go the path of features, having a static class with "well known" features would be good, at least ones we regularly expect people to depend on (like V2 source generator API, or maybe IOperation if we had it at the time, etc...)